File: c/stack/array_stack.c
/* constructor */
ArrayStack *newArrayStack() {
    ArrayStack *stack = malloc(sizeof(ArrayStack));
    // initialize enough memory to avoid segmentation fault
    stack->data = malloc(sizeof(int) * MAX_SIZE);
    stack->size = 0;
    return stack;
}

/* destructor */
void delArrayStack(ArrayStack *stack) {
    free(stack->data);
    free(stack);
}

/* get size of stack */
int size(ArrayStack *stack) {
    return stack->size;
}

/* check if stack is empty */
bool isEmpty(ArrayStack *stack) {
    return stack->size == 0;
}

/* push element into stack */
void push(ArrayStack *stack, int num) {
    if (stack->size == MAX_SIZE) {
        printf("stack is full.\n");
        return;
    }
    stack->data[stack->size] = num;
    stack->size++;
}

/* get top element of stack */
int peek(ArrayStack *stack) {
    if (stack->size == 0) {
        printf("stack is empty.\n");
        return INT_MAX;
    }
    return stack->data[stack->size - 1];
}

/* pop top element of stack */
int pop(ArrayStack *stack) {
    if (stack->size == 0) {
        printf("stack is empty.\n");
        return INT_MAX;
    }
    int val = peek(stack);
    stack->size--;
    return val;
}

File: python/stack/array_stack.py
class ArrayStack:
    """implement stack with array"""

    def __init__(self):
        """constructor"""
        self._stack: list[int] = []

    def size(self) -> int:
        """get the size of the stack"""
        return len(self._stack)

    def is_empty(self) -> bool:
        """whether the stack is empty"""
        return self._stack == []

    def push(self, item: int):
        """push element into the stack"""
        self._stack.append(item)

    def pop(self) -> int:
        """pop element out of the stack"""
        if self.is_empty():
            raise IndexError("the stack is empty")
        return self._stack.pop()

    def peek(self) -> int:
        """get the element on the top"""
        if self.is_empty():
            raise IndexError("the stack is empty")
        return self._stack[-1]

    def to_list(self) -> list[int]:
        """convert the stack to list"""
        return self._stack
